docs/API.md

```diff
@@ -104,133 +177,185 @@ gulp.task('somename', function() {
+It will be empty for anonymous function:
 
 ```js
-gulp.task('mytask', ['array', 'of', 'task', 'names'], function() {
-  // Do stuff
-});
+// foo.name === 'foo'
+function foo() {};
+
+// bar.name === ''
+var bar = function() {};
+
+// bar.name === ''
+bar.name = 'bar'
 ```
 
-**Note:** Are your tasks running before the dependencies are complete?  Make sure your dependency tasks are correctly using the async run hints: take in a callback or return a promise or event stream.
+You should either provide the task name or avoid anonymous functions.
 
 #### fn
 
-The function that performs the task's operations. Generally this takes the form of `gulp.src().pipe(someplugin())`.
+The function that performs the task's operations.
+Generally this takes the form of `return gulp.src().pipe(someplugin());`.
 
-#### Async task support
+Gulp assumes each task is asynchronous and call them with a callback parameter.
+A Task should either call the callback parameter to signal when operations
+have ended, or return a stream, promise, child process or RxJS observable.
 
-Tasks can be made asynchronous if its `fn` does one of the following:
+#### Async support
 
 ##### Accept a callback
 
-```javascript
-// run a command in a shell
-var exec = require('child_process').exec;
-gulp.task('jekyll', function(cb) {
-  // build Jekyll
-  exec('jekyll build', function(err) {
-    if (err) return cb(err); // return error
-    cb(); // finished task
-  });
+```js
+var del = require('del');
+
+gulp.task('clean', function(done) {
+  del(['.build/'], done);
 });
 ```
 
+The callback accept an optional `Error` object. If it receives an error,
+the task will fail.
+
 ##### Return a stream
 
 ```js
 gulp.task('somename', function() {
-  var stream = gulp.src('client/**/*.js')
+  return gulp.src('client/**/*.js')
     .pipe(minify())
     .pipe(gulp.dest('build'));
-  return stream;
 });
 ```
 
 ##### Return a promise
 
-```javascript
-var Q = require('q');
+```js
+var Promise = require('promise');
+var del = require('del');
+
+gulp.task('clean', function() {
+  return new Promise(function (resolve, reject) {
+    del(['.build/'], function(err) {
+      if (err) {
+        reject(err);
+      } else {
+        resolve();
+      }
+    });
+  });
+});
+```
 
-gulp.task('somename', function() {
-  var deferred = Q.defer();
+or:
+```js
+var promisedDel = require('promised-del');
 
-  // do async stuff
-  setTimeout(function() {
-    deferred.resolve();
-  }, 1);
+gulp.task('clean', function() {
+  return promisedDel(['.build/']);
+});
+```
+
+##### Return a child process
 
-  return deferred.promise;
+```js
+gulp.task('clean', function() {
+  return spawn('rm', ['-rf', path.join(__dirname, 'build')]);
 });
+
 ```
 
-**Note:** By default, tasks run with maximum concurrency -- e.g. it launches all the tasks at once and waits for nothing. If you want to create a series where tasks run in a particular order, you need to do two things:
+##### Return a RxJS observable
```

Comment:
Is it work putting a link to RxJS here somewhere? Most people haven't heard of it.